File: iOS/BCFerries/FerryRepository.swift
//
//  FerryRepository.swift
//  BCFerries
//

import Foundation

open class FerryRepository {
    private let ferryDao = FerryDao()

    open func getFerries(from: FerryPier, dow: FerryDay) -> [Ferry] {
        return ferryDao.query(from: from, dow: dow)
    }

    open func shouldRefresh() -> Bool {
        return false
    }

    open func refresh(completion: @escaping () -> Void) {
        BCFerriesApiFetcher.fetch { ferries, _ in
            if ferries.count >= 40 {
                self.ferryDao.save(result: ferries, piers: FerryPier.ENUMS)
            } else {
                // Offline or the feed is down: fall back to the schedule bundled with the app
                JsonFetcher.fetch { ferries, _ in
                    if ferries.count >= 40 {
                        self.ferryDao.save(result: ferries, piers: FerryPier.ENUMS)
                    }
                }
            }
            completion()
        }
    }

}

File: Android/app/src/main/java/com/lunesu/bcferries/FerryRepository.kt
package com.lunesu.bcferries

import android.util.Log

open class FerryRepository(db: DbOpenHelper) {
    companion object {
        private const val TAG = "FerryRepository"
    }

    private val ferryDao = FerryDao(db)

    open fun getFerries(from: FerryPier, dow: FerryDay): List<Ferry> {
        return ferryDao.query(from, dow)
    }

    open fun shouldRefresh(): Boolean {
        return false
    }

    open suspend fun refresh() {
        runCatching {
            ferryDao.save(Utils.atLeast(BCFerriesApiFetcher.fetch(), 40), *FerryPier.ENUMS)
        }.onFailure {
            // Keep whatever schedule we already have
            Log.w(TAG, "refresh failed", it)
        }
    }
}
